### ISSUE TITLE
Progress Anomaly Maps

### ISSUE THREAD

**Issue by KJ-Waller** [Carmoondedraak/FACT-2021#7]
MNIST Results:
We used a standard variational autoencoder. We trained with inlier digits 1 and 3. Each model was trained with batch size of 128, where each image is an unnormalized grayscale image of 28x28x1, training for a total of 150 epochs each. Each model was trained 3 times, with 3 different learning rates: 0.001, 0.0005, and 0.0001

Loss curves for the 3 models trained on digit 1:
![mnist_loss_digit1](https://user-images.githubusercontent.com/28184973/105477848-b66d3600-5ca2-11eb-80cf-a60e16ed933b.png)
Attention maps for the three models (trained with different learning rates) evaluated on digits 7.
![mnist_attmaps_d1d7_lr1e-3](https://user-images.githubusercontent.com/28184973/105477927-d3096e00-5ca2-11eb-90a1-37facb2eef81.png) ![mnist_attmaps_d1d7_lr1e-4](https://user-images.githubusercontent.com/28184973/105477934-d56bc800-5ca2-11eb-9303-b6dc4feed5be.png) ![mnist_attmaps_d1d7_lr5e-4](https://user-images.githubusercontent.com/28184973/105477941-d69cf500-5ca2-11eb-8526-108749cf64bf.png)

Loss curves for the 3 models trained on digit 3:
![mnist_loss_digit3](https://user-images.githubusercontent.com/28184973/105478026-ee747900-5ca2-11eb-82d4-509c0c8c35ee.png)
Attention maps for the three models (trained with different learning rates) evaluated on digits 7.
![mnist_attmaps_d3d7_lr1e-3](https://user-images.githubusercontent.com/28184973/105478042-f502f080-5ca2-11eb-8cb9-e06219d2e352.png) ![mnist_attmaps_d3d7_lr1e-4](https://user-images.githubusercontent.com/28184973/105478050-f7fde100-5ca2-11eb-85b0-65a7e9eff427.png) ![mnist_attmaps_d3d7_lr5e-4](https://user-images.githubusercontent.com/28184973/105478057-f9c7a480-5ca2-11eb-952b-52873dab6fae.png)

UPDATE:
The best performing models for MNIST in terms of VAE loss were trained with a learning rate of 0.0005, which are model number 2 for digit 1 and model number 5 for digit 3. Below we show some of the attention maps of these two models evaluated on outlier digits.

For model 2 trained on digit 1, evaluated on digits 7, 4, 9 and 2, as in the paper:
![batch4-attmaps](https://user-images.githubusercontent.com/28184973/105630025-c3c32580-5e46-11eb-9f18-5ea5d6b3c211.png) ![batch6-attmaps](https://user-images.githubusercontent.com/28184973/105630029-cd4c8d80-5e46-11eb-8474-c7fe6a0c2d0c.png) ![batch8-attmaps](https://user-images.githubusercontent.com/28184973/105630036-d76e8c00-5e46-11eb-9017-03ec2e25a060.png) ![batch23-attmaps](https://user-images.githubusercontent.com/28184973/105630041-de959a00-5e46-11eb-97fd-e34d8d05caea.png)

For model 5 trained on digit 3, evaluated on digits 8 and 5:
![batch21-attmaps](https://user-images.githubusercontent.com/28184973/105630095-4c41c600-5e47-11eb-8d7f-bb32e6f0d0a3.png) ![batch14-attmaps](https://user-images.githubusercontent.com/28184973/105630101-56fc5b00-5e47-11eb-88b2-f3730c7590c8.png)

**Comment by KJ-Waller** [comment]
UCSD Results:
We extended the variational autoencoder used on MNIST, by adding two convolutional layers to deal with images of 100x100x1 from the UCSD dataset. Again, no normalization was performed on the images. We trained for a total of 3 models with 3 different learning rates: 0.001, 0.0005 and 0.0001. The models were trained with a batch size of 128. For each model, we evaluated them by visualizing the generated attention maps by using the activations and gradients from 3 different layers, as mentioned in the paper, with output size 50x50, 25x25 and 12x12.

Loss curves for the 3 models:
![ucsd_loss](https://user-images.githubusercontent.com/28184973/105478750-df41fb00-5ca3-11eb-9a7f-02416b114875.png)

Each of the models was quantitatively evaluated on a test set containing images with pedestrians and vehicles, by calculating the AUROC of the attention maps with as targets the masks. Coincidentally, each model which was trained with different learning rates, achieved the same AUROC regardless of which convolutional layer was used for attention map generation, so we only show the AUROC of the 3 models separately.

- Model 1 w/ lr 0.001 AUROC: 0.5002
- Model 2 w/ lr 0.0005 AUROC: 0.6533
- Model 3 w/ lr 0.0001 AUROC: 0.6689

The following results are the input images, attention maps, generated binary localization images and the target masks for model 3 with the best AUROC.
![batch30-input](https://user-images.githubusercontent.com/28184973/105479792-2f6d8d00-5ca5-11eb-8a11-d75a0be3c5b1.png) ![batch30-attmaps](https://user-images.githubusercontent.com/28184973/105479797-31cfe700-5ca5-11eb-8f36-c5c7a2bece0f.png) ![batch30-blocmaps](https://user-images.githubusercontent.com/28184973/105479804-34cad780-5ca5-11eb-884d-bbc3109de36a.png) ![batch30-targets](https://user-images.githubusercontent.com/28184973/105479808-36949b00-5ca5-11eb-990c-0f22f52f8ad4.png)

**Comment by KJ-Waller** [comment]
MVTEC Results:
We report some preliminary results on the MVTEC dataset. For this dataset, we used a Resnet-18 based VAE on images of 256x256x3. The images were normalized in this case. One of the better performing models in terms of the AUROC was trained on images of hazelnuts in the MVTEC dataset. This model had a learning rate of 0.0005, and achieved a AUROC 0.8225. Due to the size of the model and images, the batch size had to be restricted to 16 in order to fit it into VRAM. The model is yet to be evaluated on generating the attention maps from different layers.

Below are some results for a batch of 16 hazelnuts. We show the original input image, the generated attention maps, the generated binary localization maps and the target masks.
![batch3-input](https://user-images.githubusercontent.com/28184973/105484727-19af9600-5cac-11eb-9c5a-ac595411cb68.png)
![batch3-attmaps](https://user-images.githubusercontent.com/28184973/105484745-2207d100-5cac-11eb-89fd-c45518c3e9e3.png)
![batch3-blocmaps](https://user-images.githubusercontent.com/28184973/105484773-2b913900-5cac-11eb-9a0a-41f82fbf56e9.png)
![batch3-targets](https://user-images.githubusercontent.com/28184973/105484778-2df39300-5cac-11eb-9bba-b21a7908d9ab.png)

We also include some reconstructed images for a batch, which do show a great lack of detail.
![480-rec](https://user-images.githubusercontent.com/28184973/105484896-5e3b3180-5cac-11eb-8d68-fd65531afe3a.png)

UPDATE:
| MVTEC Object | Results | Seed |
| ---------------------|-----------|---------|
| Carpet | 0.68| 3|
| | 0.56| |
| Grid | 0.69| 3 |
| | 0.52| |
| Leather | 0.80| 3 |
| | 0.55| |
| Tile | 0.73| 3 |
| | 0.54| |
| Wood | 0.7| 3 |
| | 0.54| |
| Bottle | 0.7| 3 |
| | 0.59| |
| Cable | 0.79| 3 |
| | 0.59| |
| Capsule | 0.82| 3 |
| | 0.50| |
| Hazelnut | 0.9| 3 |
| | 0.6| |
| Metal Nut | 0.8| 3 |
| | 0.5| |
| Pill | 0.9| 2 |
| | 0.55| |
| Screw | 0.91| 3 |
| | 0.48| |
| Toothbrush | 0.83| 3 |
| | 0.49| |
| Transistor | 0.72| 3 |
| | 0.48| |
| Zipper | 0.67| 3 |
| | 0.48| |

Below, for each MVTec-AD object, we show broken objects, the attention maps, the generated binary localization maps and target masks:
Carpet:
![batch2-input](https://user-images.githubusercontent.com/28184973/105712187-c0e53500-5f19-11eb-9fb0-bf0d88bc208f.png) ![batch2-attmaps](https://user-images.githubusercontent.com/28184973/105712196-c478bc00-5f19-11eb-9c79-bc4bc2fd4147.png) ![batch2-blocmaps](https://user-images.githubusercontent.com/28184973/105712236-cfcbe780-5f19-11eb-890e-4d7ce5d5db2e.png) ![batch2-targets](https://user-images.githubusercontent.com/28184973/105712243-d22e4180-5f19-11eb-97f8-cec78a9f3ba2.png)

Grid:
![batch13-input](https://user-images.githubusercontent.com/28184973/105714503-baa48800-5f1c-11eb-92db-396207e82bd4.png)
![batch13-attmaps](https://user-images.githubusercontent.com/28184973/105714514-bd06e200-5f1c-11eb-846b-b0991e9528fa.png)
![batch13-blocmaps](https://user-images.githubusercontent.com/28184973/105714530-c001d280-5f1c-11eb-9957-25ff247dba05.png)
![batch13-targets](https://user-images.githubusercontent.com/28184973/105714538-c1cb9600-5f1c-11eb-8337-46838d3f4d71.png)

Leather:
![batch5-input](https://user-images.githubusercontent.com/28184973/105714707-fb9c9c80-5f1c-11eb-90b9-a5fcecf5054f.png)
![batch5-attmaps](https://user-images.githubusercontent.com/28184973/105714711-fd666000-5f1c-11eb-9a6d-eafeace81b28.png)
![batch5-blocmaps](https://user-images.githubusercontent.com/28184973/105714714-ff302380-5f1c-11eb-8aeb-977677e4fab4.png)
![batch5-targets](https://user-images.githubusercontent.com/28184973/105714719-00615080-5f1d-11eb-9096-8f8f2a1b66e3.png)

Tile:
![batch5-input](https://user-images.githubusercontent.com/28184973/105713115-03f3d800-5f1b-11eb-988b-b53386c6cd20.png)
![batch5-attmaps](https://user-images.githubusercontent.com/28184973/105713128-081ff580-5f1b-11eb-9c59-c86df1f68adf.png)
![batch5-blocmaps](https://user-images.githubusercontent.com/28184973/105713140-0a824f80-5f1b-11eb-98ab-4dad85d6798c.png)
![batch5-targets](https://user-images.githubusercontent.com/28184973/105713146-0c4c1300-5f1b-11eb-8f41-9227f0e69edf.png)

Wood:
![batch2-input](https://user-images.githubusercontent.com/28184973/105713584-9dbb8500-5f1b-11eb-8fce-73b1e312c821.png)
![batch2-attmaps](https://user-images.githubusercontent.com/28184973/105713593-a0b67580-5f1b-11eb-8489-b3c6956dd5b0.png)
![batch2-blocmaps](https://user-images.githubusercontent.com/28184973/105713600-a1e7a280-5f1b-11eb-8943-1c3de7c834b3.png)
![batch2-targets](https://user-images.githubusercontent.com/28184973/105713602-a3b16600-5f1b-11eb-9d46-cd74dee7a64f.png)

Bottle:
![batch10-input](https://user-images.githubusercontent.com/28184973/105714790-1707a780-5f1d-11eb-911c-2ace73db2621.png)
![batch10-attmaps](https://user-images.githubusercontent.com/28184973/105714800-1a9b2e80-5f1d-11eb-884d-2c4a3f4b183a.png)
![batch10-blocmaps](https://user-images.githubusercontent.com/28184973/105714820-1f5fe280-5f1d-11eb-8170-ee1cf9606789.png)
![batch10-targets](https://user-images.githubusercontent.com/28184973/105714829-2129a600-5f1d-11eb-91e6-4a9bdc47d4f3.png)

Cable:
![batch15-input](https://user-images.githubusercontent.com/28184973/105713977-0c004780-5f1c-11eb-973a-0ec8d5021ee0.png)
![batch15-attmaps](https://user-images.githubusercontent.com/28184973/105713986-0efb3800-5f1c-11eb-9157-22721735fa6e.png)
![batch15-blocmaps](https://user-images.githubusercontent.com/28184973/105713995-11f62880-5f1c-11eb-9a6a-07efcc064040.png)
![batch15-targets](https://user-images.githubusercontent.com/28184973/105713998-13275580-5f1c-11eb-85bd-813d2a0df649.png)

Capsule:
![batch5-input](https://user-images.githubusercontent.com/28184973/105714192-55509700-5f1c-11eb-8cfa-d6f8e83d28ed.png)
![batch5-attmaps](https://user-images.githubusercontent.com/28184973/105714201-57b2f100-5f1c-11eb-8c0f-e36bdc2a93a8.png)
![batch5-blocmaps](https://user-images.githubusercontent.com/28184973/105714209-5a154b00-5f1c-11eb-88ed-641de440954a.png)
![batch5-targets](https://user-images.githubusercontent.com/28184973/105714217-5bdf0e80-5f1c-11eb-8509-0c8999484253.png)

Hazelnut:
![batch2-input](https://user-images.githubusercontent.com/28184973/105714606-dc9e0a80-5f1c-11eb-9054-832d79e5682c.png)
![batch2-attmaps](https://user-images.githubusercontent.com/28184973/105714621-ddcf3780-5f1c-11eb-891b-e16cdfa2fba1.png)
![batch2-blocmaps](https://user-images.githubusercontent.com/28184973/105714625-df98fb00-5f1c-11eb-9d78-1f4a987619c9.png)
![batch2-targets](https://user-images.githubusercontent.com/28184973/105714631-e162be80-5f1c-11eb-9bdd-2607f63dc92a.png)

Metal Nut:
![batch15-input](https://user-images.githubusercontent.com/28184973/105712624-5a144b80-5f1a-11eb-99c7-763c5547b95d.png) ![batch15-attmaps](https://user-images.githubusercontent.com/28184973/105712653-65677700-5f1a-11eb-8a28-659bfa2c5cfb.png) ![batch15-blocmaps](https://user-images.githubusercontent.com/28184973/105712663-67c9d100-5f1a-11eb-94ff-f93a1a248440.png) ![batch15-targets](https://user-images.githubusercontent.com/28184973/105712671-6a2c2b00-5f1a-11eb-9c45-3b19f6901922.png)

Pill:
![batch1-input](https://user-images.githubusercontent.com/28184973/105712779-95167f00-5f1a-11eb-9493-0c192b528f30.png)
![batch1-attmaps](https://user-images.githubusercontent.com/28184973/105712786-99429c80-5f1a-11eb-9d34-44de2d610fab.png)
![batch1-blocmaps](https://user-images.githubusercontent.com/28184973/105712806-9e075080-5f1a-11eb-9bf2-c1740a7346e3.png)
![batch1-targets](https://user-images.githubusercontent.com/28184973/105712815-a2336e00-5f1a-11eb-9b68-54c3bf7f6d76.png)

Screw:
![batch1-input](https://user-images.githubusercontent.com/28184973/105712923-cbec9500-5f1a-11eb-9f53-cd10b8e854cb.png)
![batch1-attmaps](https://user-images.githubusercontent.com/28184973/105712938-cee78580-5f1a-11eb-95b9-bf3840dfb3a2.png)
![batch1-blocmaps](https://user-images.githubusercontent.com/28184973/105712961-d73fc080-5f1a-11eb-9e70-d3a2aac37863.png)
![batch1-targets](https://user-images.githubusercontent.com/28184973/105712970-d9a21a80-5f1a-11eb-9657-3950e5b7f4ad.png)

Toothbrush:
![batch3-input](https://user-images.githubusercontent.com/28184973/105713248-300f5900-5f1b-11eb-8549-633b89ff3b1e.png)
![batch3-attmaps](https://user-images.githubusercontent.com/28184973/105713253-3271b300-5f1b-11eb-9aa3-bd1028a4542e.png)
![batch3-blocmaps](https://user-images.githubusercontent.com/28184973/105713259-34d40d00-5f1b-11eb-9fcb-cd6f3298929b.png)
![batch3-targets](https://user-images.githubusercontent.com/28184973/105713264-36053a00-5f1b-11eb-93b2-34ddfd5848df.png)

Transistor:
![batch9-input](https://user-images.githubusercontent.com/28184973/105713471-7f558980-5f1b-11eb-9538-3f2bd0c2ac1a.png)
![batch9-attmaps](https://user-images.githubusercontent.com/28184973/105713531-8c727880-5f1b-11eb-9be1-84a7c4e5de66.png)
![batch9-blocmaps](https://user-images.githubusercontent.com/28184973/105713535-8f6d6900-5f1b-11eb-9ec9-a87a9ea5c24c.png)
![batch9-targets](https://user-images.githubusercontent.com/28184973/105713538-909e9600-5f1b-11eb-9a46-ed5398934bb4.png)

Zipper:
![batch9-input](https://user-images.githubusercontent.com/28184973/105713757-c8a5d900-5f1b-11eb-8a96-f89b2e6b467b.png)
![batch9-attmaps](https://user-images.githubusercontent.com/28184973/105713765-cba0c980-5f1b-11eb-9754-7891297c96c9.png)
![batch9-blocmaps](https://user-images.githubusercontent.com/28184973/105713772-cd6a8d00-5f1b-11eb-90e0-167159ec79e2.png)
![batch9-targets](https://user-images.githubusercontent.com/28184973/105713778-ce9bba00-5f1b-11eb-8073-0277221c682c.png)
